Implementation of 'test' project support and sub-namespaces

goldhoorn opened the pull request:
Test-projects still not compiling successfully

goldhoorn commented:
I currently got the following error on my testproject:

```
/home/goldhoorn/dev-avalon-21/avalon/orogen/sysmon/tests/.orogen/typekit/transports/typelib/__test_sysmon__TestTaskFoo_STATES.cpp:7:142: error: 'orogen_transports::TypelibMarshallerBase* orogen_typekits::test_sysmon_TestTaskFoo_STATES_TypelibMarshaller(const Typelib::Registry&)' should have been declared inside 'orogen_typekits'
ICECC[4754] 14:32:33: Compiled on 10.250.3.187
tests/.orogen/typekit/transports/typelib/CMakeFiles/test_sysmon-transport-typelib-gnulinux.dir/build.make:54: recipe for target 'tests/.orogen/typekit/transports/typelib/CMakeFiles/test_sysmon-transport-typelib-gnulinux.dir/__test_sysmon__TestTaskFoo_STATES.cpp.o' failed
make[2]: *** [tests/.orogen/typekit/transports/typelib/CMakeFiles/test_sysmon-transport-typelib-gnulinux.dir/__test_sysmon__TestTaskFoo_STATES.cpp.o] Error 1
make[2]: Leaving directory '/home/goldhoorn/dev-avalon-21/avalon/orogen/sysmon/build'
CMakeFiles/Makefile2:800: recipe for target 'tests/.orogen/typekit/transports/typelib/CMakeFiles/test_sysmon-transport-typelib-gnulinux.dir/all' failed
make[1]: *** [tests/.orogen/typekit/transports/typelib/CMakeFiles/test_sysmon-transport-typelib-gnulinux.dir/all] Error 2
make[1]: *** Warte auf noch nicht beendete Prozesse...
/home/goldhoorn/dev-avalon-21/avalon/orogen/sysmon/tests/.orogen/typekit/transports/corba/TransportPlugin.cpp:11:23: error: 'orogen_typekits::test_sysmonCorbaTransportPlugin' has not been declared
/home/goldhoorn/dev-avalon-21/avalon/orogen/sysmon/tests/.orogen/typekit/transports/corba/TransportPlugin.cpp: In function 'bool registerTransport(std::string, RTT::types::TypeInfo*)':
/home/goldhoorn/dev-avalon-21/avalon/orogen/sysmon/tests/.orogen/typekit/transports/corba/TransportPlugin.cpp:20:59: error: 'test_sysmon_TestTaskFoo_STATES_CorbaTransport' was not declared in this scope
/home/goldhoorn/dev-avalon-21/avalon/orogen/sysmon/tests/.orogen/typekit/transports/corba/TransportPlugin.cpp: At global scope:
/home/goldhoorn/dev-avalon-21/avalon/orogen/sysmon/tests/.orogen/typekit/transports/corba/TransportPlugin.cpp:25:30: error: 'orogen_typekits::test_sysmonCorbaTransportPlugin' has not been declared
/home/goldhoorn/dev-avalon-21/avalon/orogen/sysmon/tests/.orogen/typekit/transports/corba/TransportPlugin.cpp:25:82: error: non-member function 'std::string getTransportName()' cannot have cv-qualifier
/home/goldhoorn/dev-avalon-21/avalon/orogen/sysmon/tests/.orogen/typekit/transports/corba/TransportPlugin.cpp:27:30: error: 'orogen_typekits::test_sysmonCorbaTransportPlugin' has not been declared
/home/goldhoorn/dev-avalon-21/avalon/orogen/sysmon/tests/.orogen/typekit/transports/corba/TransportPlugin.cpp:27:80: error: non-member function 'std::string getTypekitName()' cannot have cv-qualifier
/home/goldhoorn/dev-avalon-21/avalon/orogen/sysmon/tests/.orogen/typekit/transports/corba/TransportPlugin.cpp:29:30: error: 'orogen_typekits::test_sysmonCorbaTransportPlugin' has not been declared
/home/goldhoorn/dev-avalon-21/avalon/orogen/sysmon/tests/.orogen/typekit/transports/corba/TransportPlugin.cpp:29:73: error: non-member function 'std::string getName()' cannot have cv-qualifier
/home/goldhoorn/dev-avalon-21/avalon/orogen/sysmon/tests/.orogen/typekit/transports/corba/TransportPlugin.cpp: In function 'bool loadRTTPlugin(RTT::TaskContext*)':
/home/goldhoorn/dev-avalon-21/avalon/orogen/sysmon/tests/.orogen/typekit/transports/corba/TransportPlugin.cpp:32:233: error: expected type-specifier
/home/goldhoorn/dev-avalon-21/avalon/orogen/sysmon/tests/.orogen/typekit/transports/corba/TransportPlugin.cpp: In function 'std::string getRTTPluginName()':
/home/goldhoorn/dev-avalon-21/avalon/orogen/sysmon/tests/.orogen/typekit/transports/corba/TransportPlugin.cpp:32:422: error: 'test_sysmonCorbaTransportPlugin' is not a member of 'orogen_typekits'
/home/goldhoorn/dev-avalon-21/avalon/orogen/sysmon/tests/.orogen/typekit/transports/corba/TransportPlugin.cpp:32:482: error: 'tk' was not declared in this scope
/home/goldhoorn/dev-avalon-21/avalon/orogen/sysmon/tests/.orogen/typekit/transports/corba/TransportPlugin.cpp:32:496: warning: control reaches end of non-void function [-Wreturn-type]
ICECC[4759] 14:32:33: Compiled on 10.250.3.171
tests/.orogen/typekit/transports/corba/CMakeFiles/test_sysmon-transport-corba-gnulinux.dir/build.make:84: recipe for target 'tests/.orogen/typekit/transports/corba/CMakeFiles/test_sysmon-transport-corba-gnulinux.dir/TransportPlugin.cpp.o' failed
make[2]: *** [tests/.orogen/typekit/transports/corba/CMakeFiles/test_sysmon-transport-corba-gnulinux.dir/TransportPlugin.cpp.o] Error 1
make[2]: Leaving directory '/home/goldhoorn/dev-avalon-21/avalon/orogen/sysmon/build'
CMakeFiles/Makefile2:741: recipe for target 'tests/.orogen/typekit/transports/corba/CMakeFiles/test_sysmon-transport-corba-gnulinux.dir/all' failed
make[1]: *** [tests/.orogen/typekit/transports/corba/CMakeFiles/test_sysmon-transport-corba-gnulinux.dir/all] Error 2
make[1]: Leaving directory '/home/goldhoorn/dev-avalon-21/avalon/orogen/sysmon/build'
Makefile:107: recipe for target 'all' failed
make: *** [all] Error 2
make: Leaving directory '/home/goldhoorn/dev-avalon-21/avalon/orogen/sysmon/build'
```

@doudou  need some hints here, also for the rought direction (i expecting remarks ;) )

doudou commented:
OK ... I really did not think about it this way. 

While it's not a bad idea per se, the (bad) state of the code generator and of the CMake code *will* make it difficult. I.e. there's probably globals and stuff like that which breaks generating two projects within the same ruby instance.  The cleanup would be welcome, but would be quite a bit of work.

I initially thought the idea was to have categories in the orogen project (i.e. something like)

~~~
with 'tests' do
   task_context 'BlaBla' do
   end
end
~~~

which would be enabled/disabled by the command line. The test task contexts would then be built only if requested.

goldhoorn commented:
Then the test-tasks would be also inside the tasks/ directory
Otherwise this would ger harder because we should go for something like 
test_task_context 'bla' do and load this from tests/ instead of tasks/ 
but this would cause a wider separation.

Another idea i had is to completly ignore (from a orogen-point of view) 
the test-folder and generate directlry from autobuild a sub-test-package.

But i would prefer a cleanup for orogen so that within the same instance 
the test can be build (or not).

Thoughts?

Matthias

P.S. my approach enables/disables the generation from commandline too

On 15.04.2015 15:15, Sylvain Joyeux wrote:
>
> OK ... I really did not think about it this way.
>
> While it's not a bad idea per se, the (bad) state of the code 
> generator and of the CMake code /will/ make it difficult. I.e. there's 
> probably globals and stuff like that which breaks generating two 
> projects within the same ruby instance. The cleanup would be welcome, 
> but would be quite a bit of work.
>
> I initially thought the idea was to have categories in the orogen 
> project (i.e. something like)
>
> |with 'tests' do
>     task_context 'BlaBla' do
>     end
> end
> |
>
> which would be enabled/disabled by the command line. The test task 
> contexts would then be built only if requested.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/orocos-toolchain/orogen/pull/38#issuecomment-93389817>.
>


-- 
  Dipl.-Inf. Matthias Goldhoorn
  Space and Underwater Robotic

  Universität Bremen
  FB 3 - Mathematik und Informatik
  AG Robotik
  Robert-Hooke-Straße 1
  28359 Bremen, Germany
  
  Zentrale: +49 421 178 45-6611
  
  Besuchsadresse der Nebengeschäftstelle:
  Robert-Hooke-Straße 5
  28359 Bremen, Germany
  
  Tel.:    +49 421 178 45-4193
  Empfang: +49 421 178 45-6600
  Fax:     +49 421 178 45-4150
  E-Mail:  matthias.goldhoorn@informatik.uni-bremen.de

  Weitere Informationen: http://www.informatik.uni-bremen.de/robotik

doudou commented:
> Then the test-tasks would be also inside the tasks/ directory

Would that be so much of a problem ? Another issue is that if tests/ is used for the C++ code, it would feel awkward for me to have the actual unit tests in there (which are in Ruby)

> Another idea i had is to completly ignore (from a orogen-point of view) the test-folder and generate directlry from autobuild a sub-test-package.

I thought about that as well, but I don't like it for the same reasons that you don't

> P.S. my approach enables/disables the generation from commandline too

I noticed ;-)

goldhoorn commented:
On 15.04.2015 15:47, Sylvain Joyeux wrote:
>
>     Then the test-tasks would be also inside the tasks/ directory
>
> Would that be so much of a problem ? Another issue is that if tests/ 
> is used for the C++ code, it would feel awkward for me to have the 
> actual unit tests in there (which are in Ruby)
>
I don't want to crowd the tasks/ folder with tests, maybe i can accept 
something like tasks/tests/ ?
But i still would prefer the tests/ folder. We can then use something 
like tests/run_suite.rb as default entry



-- 
  Dipl.-Inf. Matthias Goldhoorn
  Space and Underwater Robotic

  Universität Bremen
  FB 3 - Mathematik und Informatik
  AG Robotik
  Robert-Hooke-Straße 1
  28359 Bremen, Germany
  
  Zentrale: +49 421 178 45-6611
  
  Besuchsadresse der Nebengeschäftstelle:
  Robert-Hooke-Straße 5
  28359 Bremen, Germany
  
  Tel.:    +49 421 178 45-4193
  Empfang: +49 421 178 45-6600
  Fax:     +49 421 178 45-4150
  E-Mail:  matthias.goldhoorn@informatik.uni-bremen.de

  Weitere Informationen: http://www.informatik.uni-bremen.de/robotik

doudou commented:
> I don't want to crowd the tasks/ folder with tests, maybe i can accept something like tasks/tests/ ?

I would have that rather than putting the tasks in test/ 

One other way would be to allow for a subnamespace in tasks (i.e. tests::Task). When converting names to filenames, the namespace would become a directory. That would be a pretty generic mechanism to organize tasks (and would fit your needs). I am mentioning this as I believe it might work out of the box on the C++ side. Only the CMake code would have to be modified to have proper classname-to-filename conversion. I wonder about what kind of unintented consequences it could have to the rest of the toolchain, though, honestly (so ... potentially a lot of work)

goldhoorn commented:
I updated the PR, but there is one point still not working

the gernarted ``` tasks/test/Fusel.hpp ``` wants to include:
```
#include "sysmon/Test/FuselBase.hpp"
```
but this is not found .

i cannot figure out how this (in the non-subspaced position) is working:

``` #include "sysmon/TaskBase.hpp" ```

goldhoorn commented:
*request opinions before cleanup*
open questions
1.  INCLUDE problem from above post
2. where to put flag test_support?

doudou commented on lib/orogen/gen/imports.rb:
```diff
@@ -283,6 +283,12 @@ def task_context(name, &block) # :nodoc:
+            
+            def test_task_context(name, &block)
+                if test_support
+                    task_context("Test::" + name, &block)
```
So far we use snake_case for namespaces. I would really prefer 'test' without a uppercase T

doudou commented:
>  INCLUDE problem from above post

The task classes are symlinked in .orogen/<project name>. You probably need to modify that code to create the symlink under the subdirectory.

> where to put flag test_support?

Now that I see it, I don't really like the test_task_context stanza. It hides the fact that the task is under the test:: namespace, which makes it hard to follow.

Two options to go forward:
1. a command-line option to exclude some namespace from the generation (i.e. --exclude=test::)
2. a block form for tests

~~~
with :test do
   task_context 'test::Task' do
   end
end
~~~

I would prefer (1) personally.

goldhoorn commented:
updated the PR
Fixed:
1. I keep the test_task_context but made it only a wrapper
2. the symlink generation is now working correctly 
Open:
1. Where to put the excluded_namespace it is currently twice represented and i'm not know if this is needed 
2. i disabled the test namespace by default (i prefer)
3. (cleanup)

doudou commented on bin/orogen:
```diff
@@ -154,26 +157,33 @@ end
-                project.using_typekit tk_name
-            rescue LoadError
-                raise ArgumentError, "#{tk_name}, which was given to the --import option, does not seem to be a typekit name"
+    f = [[File.expand_path(files.first),project]]
```
This part should go away now, no ?

doudou commented on bin/orogen:
```diff
@@ -80,7 +82,8 @@ parser = OptionParser.new do |opt|
     end
-
+    
+    opt.on("-t", "--tests")   { rtt_cpp.enable_namespace = "test"}
```
I would place the enable_namespace / disable_namespace calls directly on the project object (no global)

doudou replied:
You seem to assume that there will be only one namespace ... Should replace the '=' by a plain method call

~~~
project.enable_namespace 'test'
~~~

doudou commented on lib/orogen/gen/project.rb:
```diff
@@ -101,12 +106,17 @@ class Project
                 RTT_CPP.orocos_target.dup
             end
+        
+            def enable_namespace=(value); RTT_CPP.enable_namespace = value end
```
This is absolutely horrible. You make a local method change a global state.

doudou commented on lib/orogen/gen/project.rb:
```diff
@@ -48,11 +48,16 @@ class << self
         end
-
+           
+        def self.enable_namespace=(value); @disabled_namespaces.delete(value) end
```
This is why some people hate operator overloading. You make assignation '=' behave resp. like << and delete. Yuk.

doudou commented on lib/orogen/spec/project.rb:
```diff
@@ -136,6 +144,11 @@ def default_task_superclass
             # the documentation of that class for more details.
 	    def task_context(name, options = Hash.new, &block)
+                if namespace_disabled?(name.split("::")[0..-2].join("::"))
+                    OroGen::Gen::RTT_CPP.warn ("Task #{name} is disabled, because the namespace " + name.split("::")[0..-2].join("::") + " is disabled")
```
This should *DEFINITELY* not be a warning. This will be a perfectly common and normal situtation !

doudou commented:
One thing you overlooked: you need to generate enable_namespace / disable_namespace stanzas in the installed orogen file. Otherwise, the setting will not be propagated when we use the orogen spec (after the project got installed). See templates/project.orogen

goldhoorn commented:
On 17.04.2015 15:03, Sylvain Joyeux wrote:
>
> One thing you overlooked: you need to generate enable_namespace / 
> disable_namespace stanzas in the installed orogen file. Otherwise, the 
> setting will not be propagated when we use the orogen spec (after the 
> project got installed). See templates/project.orogen
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/orocos-toolchain/orogen/pull/38#issuecomment-93981541>.
>

This was the important part i missed between the gen/spec part, Thanks.
I had this in mind already but could not assign this.

I fixed this generation no and done the code/commit cleanup.

-- 
  Dipl.-Inf. Matthias Goldhoorn
  Space and Underwater Robotic

  Universität Bremen
  FB 3 - Mathematik und Informatik
  AG Robotik
  Robert-Hooke-Straße 1
  28359 Bremen, Germany
  
  Zentrale: +49 421 178 45-6611
  
  Besuchsadresse der Nebengeschäftstelle:
  Robert-Hooke-Straße 5
  28359 Bremen, Germany
  
  Tel.:    +49 421 178 45-4193
  Empfang: +49 421 178 45-6600
  Fax:     +49 421 178 45-4150
  E-Mail:  matthias.goldhoorn@informatik.uni-bremen.de

  Weitere Informationen: http://www.informatik.uni-bremen.de/robotik

doudou commented on lib/orogen/gen/tasks.rb:
```diff
@@ -300,20 +300,40 @@ def register_for_generation
+            # The name of the header file containing the C++ code which defines
+            # this task context
+            def header_file
+                File.join(project.name.downcase, "#{basename}.hpp")
```
This does not look right. Should include the namespace, or ?

doudou commented on lib/orogen/gen/tasks.rb:
```diff
@@ -300,20 +300,40 @@ def register_for_generation
             # Returns the name without an eventual library name
             def basename
-                library_name, name = self.name.split("::")
+                library_name, name = self.name.split("::").last
```
This does not look right either. Just return self.name.split("::").last

doudou commented on lib/orogen/spec/project.rb:
```diff
@@ -38,8 +41,13 @@ def initialize(loader)
                 @define_default_deployments = true
                 @enabled_transports = Set.new
                 @max_sizes = Hash.new
+                @disabled_namespaces = ['test']
```
That would be wrong. While it does make sense in gen, it is IMO harmful in spec. In any case, it is broken w.r.t. the code generated in orogen as you only call disable_namespace (and therefore have no way to enable the test namespace)

doudou commented on lib/orogen/spec/project.rb:
```diff
@@ -136,6 +144,10 @@ def default_task_superclass
             # the documentation of that class for more details.
 	    def task_context(name, options = Hash.new, &block)
+                if namespace_disabled?(name.split("::")[0..-2].join("::"))
+                    return nil
```
Just return, not return nil

goldhoorn commented:
*updated*

goldhoorn commented:
*pong*

doudou commented on lib/orogen/gen/project.rb:
```diff
@@ -1011,6 +1023,10 @@ def task_context(name, options = Hash.new, &block)
                 task
 	    end
 
+            def test_task_context(name, options = Hash.new, &block)
```
I would really prefer removing this. This adds to the API with IMO close to zero benefit

doudou commented on lib/orogen/gen/tasks.rb:
```diff
@@ -300,21 +300,33 @@ def register_for_generation
-                else
-                    File.join("#{project.name.downcase}", "#{basename}.hpp")
-                end
+                File.join(project.name.downcase, basepath, "#{basename}.hpp")
```
???? Not good IMO.

This is the gen part. Loading is broken, so you don't have the guarantee that #project will actually return the task context's definition project. I would rather keep the old method until a brave soul refactors gen to use the loaders.

doudou commented on lib/orogen/templates/config/Base.cmake:
```diff
@@ -1,8 +1,10 @@
 include(OrogenPkgCheckModules)
+if(NOT TARGET regen)
```
The two ifs (this one and the one about check-uptodate) are broken. Remove them (especially since, I believe, they are not needed anymore)

doudou commented on lib/orogen/templates/typekit/CMakeLists.txt:
```diff
@@ -46,12 +46,13 @@ ENDIF ( NOT CMAKE_BUILD_TYPE )
 include(OrogenPkgCheckModules)
 orogen_pkg_check_modules(OrocosRTT REQUIRED "orocos-rtt-${OROCOS_TARGET}>=1.99.4")
 
+if(NOT TARGET regen)
```
same comment about targets

goldhoorn commented:
*updated*

doudou commented on lib/orogen/gen/tasks.rb:
```diff
@@ -300,21 +300,34 @@ def register_for_generation
             def basename
-                library_name, name = self.name.split("::")
-                name || library_name
+                name = self.name.split("::").last
```
The assignation is useless here

doudou commented:
I'm not forgetting this ... I just would like to give it a try for real before I merge it.

goldhoorn commented:
*argh* updated

goldhoorn commented:
@dodou do you have had look aready?
Otherwise i would like to go for a merge before it diverges too much. (we can later fix it if needed)

doudou commented:
Yep. No time to try it out, let's just merge it
